### DISCUSSION TITLE
Best practice to react to 401s

### DISCUSSION THREAD

**Discussion by Tekpakma** [rpldy/react-uploady#442] [Q&A]
I use JWT tokens to authenticate my upload requests. The current way I react to 401s responses (which most likely occur due to expired tokens) is with a useItemError Listener, checking if the token is expired, getting a new token and then calling the useRetry() Hook with the new token.  

This mostly works, but I think that this flow isn't ideal though, because f.E. with using TUS it skips the resuming of Uploads and starts a new one. It would probably be ideal to intercept the Request/Response, and change the onError logic.

So I'm curious, is there a smarter way to handle this all together?

**Comment by yoavniran** [comment]
Hi @Tekpakma it sounds like perhaps I need to look into **retry** in the context of TUS uploads. 

If in fact the already uploaded chunks are discarded when retrying a failed upload that sounds like something that can/should be fixed on the Uploady side.

**Reply by Tekpakma** [reply]
I checked the Code and if I'm not mistaken, then "normal" SendOptions (like Authorization Headers) aren't event applied for the Resumption of Uploads. After Resumption Fails, Uploady creates a new Upload with normal behavior (headers etc.). As for my specific problem, don't really see an easy way of fixing that now, would probably need quite some code modification...

**Reply by yoavniran** [reply]
I picked two potential issues from your descriptions:

- retry doesnt resume a Tus upload (starts from scratch)
- missing data (params) for retry of Tus upload

Are these correct? Anything else?

Are there issues you identified that **arent** related to the combination of TUS + Retry flow?

Thanks.

**Reply by Tekpakma** [reply]
Yes, these are correct!

Besides that, didn't have problems, but on a side note:
- I noticed that the chunked retries option doesn't serve a great purpose for me, because you can't pass in new/updated options.

- It would also be nice to be able to use previewMethodsRef and clear a single item of UploadPreview, instead of the whole batch.

**Reply by yoavniran** [reply]
@Tekpakma 

Why would you need to provide new/updated options for a retried chunk? Please describe the use-case so I can understand better.

**Reply by Tekpakma** [reply]
My reasoning would be the Authentication header. I can predict, that a failReason would be an expired Header, and I would get a new Auth Header before retiring the chunk, otherwise there is no point in retrying it (because it will fail again). IMO such an implementation would be cleaner, because the item wouldn't even reach an error state.

**Reply by yoavniran** [reply]
@Tekpakma 
For TUS failing to resume, as far as I can tell resume works (as long as the server responds correctly)
I closed that issue (#446) and will focus on allowing resume headers to be sent.
Unless there's a scenario im not aware of and you can add more info about?

**Reply by Tekpakma** [reply]
Yes you can close it. When I encountered this behavior, I was unaware, that the headers weren't sent as expected.

**Reply by yoavniran** [reply]
Awesome feedback @Tekpakma, thanks for taking the time to report and follow up.
I've now merged several PRs based on your info.

Version 1.4.0 will soon be out with new functionality:

- provide resumeHeaders for Tus uploads (#450)
- handle Tus resume start event to either cancel or update info dynamically (#450)
- ability to remove a single preview in UploadPreview (#451)

Doc site will be updated with relevant info shortly after

**Comment by yoavniran** [comment]
[1.4.0-rc.0](https://github.com/rpldy/react-uploady/releases/tag/v1.4.0-rc.0) RC version is out now with the discussed improvements
